proposals/xxxx-work-graphs.md

```diff
@@ -0,0 +1,535 @@
+  bool FinishedCrossGroupSharing();
+};
+
+template <typename RecordTy>
+using GroupNodeInputRecord = detail::GroupNodeInputRecordBase<RecordTy, false>;
+
+template <typename RecordTy>
+using RWGroupNodeInputRecord = detail::GroupNodeInputRecordBase<RecordTy, true>;
+```
+
+Coalescing launch nodes also accept the `EmptyNodeInput` input object for cases
+without record data. The pseudo-HLSL interface for `EmptyNodeInput` is:
+
+```c++
+class EmptyNodeInput {
+  /// @brief Returns the number of records that have been coalesced into the
+  /// current thread group.
+  ///
+  /// Returns 1..._MaxCount_, where _MaxCount_ is specified by the
+  /// `[MaxRecords(_MaxCount_)]` attribute applied to the parameter declaration.
+  uint Count() const;
+};
+```
+
+##### Node Output Objects
+
+
+Node output objects are either records or empty. The following pseudo-HLSL
+defines the interfaces for the `NodeOutput` and `EmptyNodeOutput` objects:
+
+```c++
+template <typename RecordTy> class NodeOutput {
+  /// @brief Allocate a new ThreadNodeOutputRecords for this thread.
+  /// @returns A handle that collects a set of thread output records.
+  /// @param NumRecords The number of records to return for the calling thread.
+  ///
+  /// ThreadNodeOutputRecords are per-thread output records. Each thread can
+  /// produce a different number of outputs which are each unique per thread.
+  ///
+  /// Must be called in thread group uniform control flow. The value of
+  /// `NumRecords` is not required to be uniform. If the value of `NumRecords`
+  /// is `0`, the object returned is zero-sized and cannot be indexed on that
+  /// thread.
+  ThreadNodeOutputRecords<RecordTy> GetThreadNodeOutputRecords(uint NumRecords);
+
+  /// @brief Allocate GroupNodeOutputRecords for this thread group.
+  /// @returns A handle that collects a set of group node output records.
+  /// @param NumRecords The number of records to return.
+  ///
+  /// GroupNodeOutputRecords are per-group output records. The output record set
+  /// is shared across the thread group and the threads work cooperatively to
+  /// produce the output records.
+  ///
+  /// Must be called in thread group uniform control flow. The value of
+  /// `NumRecords` and `this` must be uniform across the thread group. If the
+  /// value of `NumRecords` is `0`, the object returned is zero-sized and cannot
+  /// be indexed.
+  ///
+  /// This method may not be called from _thread launch_ shaders since they do
+  /// not have a thread group.
+  GroupNodeOutputRecords<RecordTy> GetGroupNodeOutputRecords(uint NumRecords);
+
+  /// @brief Returns true if the specified output node is in the work graph.
+  bool IsValid() const;
+};
+
+class EmptyNodeOutput {
+  /// @brief Adds `Count` empty output records to the thread node outputs.
+  ///
+  /// Must be called in thread group uniform control flow.
+  void ThreadIncrementOutputCount(uint Count);
+
+  /// @brief Adds `Count` empty output records to the group node outputs.
```

Comment:
```suggestion
  /// @brief Adds `Count` empty output records to the node output, once for the
  /// group, instead of summing the value across threads.
```